flink-connectors/flink-connector-upsert-test/src/main/java/org/apache/flink/connector/upserttest/sink/BinaryFileUtil.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.upserttest.sink;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.serialization.DeserializationSchema;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Collection uf utility methods for tests using the {@link UpsertTestSink}. */
+@Internal
+public class BinaryFileUtil {
```

Comment:
I think this should be a `SimpleVersionedSerializer` since that is the main entry for these kinds of operations in the Flink code base. For reading/writing many records you can take a look at https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/core/io/SimpleVersionedSerialization.java

Reply:
I previously wanted to use the SimpleVersionedSerializer and SimpleVersionedSerialization. However implementing this interface adds the same kind of code we already have in place. Also the SimpleVersionedSerialization only takes single byte arrays as input so then we would have to write another method around that to allow both key and value as input. And finally it does not  provide a simple entrypoint for reading back all elements so I decided against that

Reply:
> I previously wanted to use the SimpleVersionedSerializer and SimpleVersionedSerialization. However implementing this interface adds the same kind of code we already have in place. 

That is exactly the idea. Of course, the interface adds the same logic but it makes understanding the code easier because other Flink developers are used to `SimpleVersionedSerializer`. 


> Also the SimpleVersionedSerialization only takes single byte arrays as input so then we would have to write another method around that to allow both key and value as input. And finally it does not provide a simple entrypoint for reading back all elements so I decided against that,

This point I do not fully understand. My idea was to have the following

```java
class UpsertFileSinkSerializer implements SimpleVersionedSerializer<Map<ByteBuffer, ByteBuffer>> {
....
}
```

and `SimpleVersionedSerialization#readVersionAndDeSerialize` or `SimpleVersionedSerialization#writeVersionAndSerialize`.  WDYT?

Reply:
Can't we just use TypeInformation and its TypeSerializer/Deserializer? Imho the user doesn't need to supply any special code and we can just use what it's already there.

Reply:
As discussed offline `SimpleVersionedSerializer` seems to be overkill for the task at hand. `TypeInformation` while nice would require us to complicate the sink by introducing an extra topology to retrieve the information. We will keep it as is for now and can later change it if the necessity arises